### PR TITLE
sql,cli: improve the handling of notices

### LINKED ISSUE #45833: cli: handle rows and error
I have a usecase for a statement which can return some information as a result, as well as an error.

In the cli, depending on the display format, these rows may or may not show up:
  - raw, record shows the rows
  - the default tabular format never shows the rows
  - csv, tsv didn't show the rows in practice (but @knz points me to #28688 so it might show *some* rows in some cases).

We should make these all consistent and always show the rows even if we do receive an error.

I tested this with the code below for a specific statement in the sql executor:
```
		res.DisableBuffering()
		for i := 1; i < 10; i++ {
			res.AddRow(ctx, tree.Datums{tree.NewDString(fmt.Sprintf("some text %d", i))})
		}
		err := fmt.Errorf("FOO BAR")
		res.SetError(err)
		return makeErrEvent(err)
```

### PR DESCRIPTION
**Pull request by knz**
Fixes #45833.

tldr:
- notices printed at end in SQL shell (requested by @RaduBerinde),
- new built-in function `crdb_internal.notice()`,
- notices can now be logged with vmodule `notices=2`,
- code simplification.

Details:

The main purpose of this patch was to delay the printing
of notices in the crdb SQL shell to the end, after the results.

The motivation for that is that if there are many (e.g. thousands) of
result rows, a notice printed at the beginning or "in the middle"
would get drowned and missed by the human observer.

This patch achieves that with a straightforward, local change to the
CLI SQL code.

To unit test the change, a way to introduce notices at an arbitrary
point of query execution was also needed. To achieve this, the patch
introduces a new SQL built-in function `crdb_internal.notice()` with
the corresponding plumbing throughout the execution layers. This makes
it possible to send an out-of-band notice in the middle of SQL
execution. The unit test for the first changes already uses this new
built-in; I also foresee that we will discover that our users will
start using that in fancy ways that we are not envisioning yet.

To troubleshoot the implementation of the above, a way to inspect
the notices inside the SQL execution layer was needed; so this
patch also ensures notices are logged when the vmodule config
includes `notices=2`.

Finally, the various changes discussed here revealed that
the notice code in the SQL execution package could be simplified, so
this patch does this too.

Release justification: Category 4: Low risk, high reward changes to existing functionality

Release note (cli change): The `sql` and `demo` client commands now
display out-of-band server notices at the end of execution.

### CODE REVIEW

**Review by otan — APPROVED** [review]

**Comment by knz** [comment]
Thanks!

bors r=otan

**Comment by knz** [comment]
bors r=otan
